examples/xgboost_penguins/penguin_pipeline_local_e2e_test.py

```diff
@@ -0,0 +1,86 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""E2E Tests for penguin_pipeline_xgboost_local."""
+
+import os
+from typing import Text
+
+import tensorflow as tf
+from tfx import v1 as tfx
+from tfx.orchestration import metadata
+
+from . import penguin_pipeline_local
+
+
+class PenguinPipelineLocalEndToEndTest(tf.test.TestCase):
+  def setUp(self):
+    super().setUp()
+    self._test_dir = os.path.join(
+        os.environ.get('TEST_UNDECLARED_OUTPUTS_DIR', self.get_temp_dir()),
+        self._testMethodName)
+    self._penguin_root = os.path.dirname(__file__)
+
+    self._pipeline_name = 'xgboost_test'
+    self._data_root = os.path.join(self._penguin_root, 'data')
+    self._module_file = os.path.join(self._penguin_root, 'utils.py')
+    self._serving_model_dir = os.path.join(self._test_dir, 'serving_model')
+    self._pipeline_root = os.path.join(self._test_dir, 'tfx', 'pipelines',
+                                       self._pipeline_name)
+    self._metadata_path = os.path.join(self._test_dir, 'tfx', 'metadata',
+                                       self._pipeline_name, 'metadata.db')
+
+  def assertExecutedOnce(self, component: Text) -> None:
+    """Check the component is executed exactly once."""
+    component_path = os.path.join(self._pipeline_root, component)
+    self.assertTrue(tfx.dsl.io.fileio.exists(component_path))
+    execution_path = os.path.join(component_path, '.system',
+                                  'executor_execution')
+    execution = tfx.dsl.io.fileio.listdir(execution_path)
+    self.assertLen(execution, 1)
+
+  def assertPipelineExecution(self) -> None:
+    self.assertExecutedOnce('CsvExampleGen')
+    self.assertExecutedOnce('ExampleValidator')
+    self.assertExecutedOnce('SchemaGen')
+    self.assertExecutedOnce('StatisticsGen')
+    self.assertExecutedOnce('Trainer')
+
+  def testPenguinPipelineLocal(self):
+    tfx.orchestration.LocalDagRunner().run(
+        penguin_pipeline_local.create_pipeline(
+            pipeline_name=self._pipeline_name,
+            pipeline_root=self._pipeline_root,
+            data_root=self._data_root,
+            module_file=self._module_file,
+            metadata_path=self._metadata_path,
+            beam_pipeline_args=[]))
+
+    self.assertTrue(tfx.dsl.io.fileio.exists(self._metadata_path))
+    expected_execution_count = 5
+    metadata_config = (
+        tfx.orchestration.metadata.sqlite_metadata_connection_config(
+            self._metadata_path))
+    with metadata.Metadata(metadata_config) as m:
+      artifact_count = len(m.store.get_artifacts())
+      execution_count = len(m.store.get_executions())
+      self.assertGreaterEqual(artifact_count, execution_count)
```

Comment:
Quick question - why is this GreaterEqual instead of just Equal?

Reply:
Each component run create 1 or more artifacts. In this case, the Trainer creates 2 artifacts: `model_run` and `model`

Reply:
Why is an explicit number not being used here? Does it vary?

Reply:
I saw this assertGreaterEqual used in the penguin pipeline so used it here. But static numbers sound good, I'll make this change.